dbms/src/Common/HashTable/HashTable.h

```diff
@@ -730,17 +744,133 @@ class HashTable :
         return iterator(this, ptr);
     }
 
-    const_iterator end() const         { return const_iterator(this, buf + grower.bufSize()); }
-    const_iterator cend() const        { return end(); }
-    iterator end()                     { return iterator(this, buf + grower.bufSize()); }
+    const_iterator end() const { return const_iterator(this, buf + grower.bufSize()); }
+    const_iterator cend() const { return end(); }
+    iterator end() { return iterator(this, buf + grower.bufSize()); }
+
+    using Position = Cell *;
+    using ConstPosition = const Cell *;
+
+    Position startPos() { return nullptr; }
+    ConstPosition startPos() const { return nullptr; }
+
+    template <typename TSelf, typename Func, typename TPosition>
+    static bool forEachCell(TSelf & self, Func && func, TPosition & pos)
+    {
+        using TCell = decltype(*std::declval<TSelf>().buf);
+        using TKey = decltype(std::declval<Cell>().getKey());
+        static constexpr bool with_key = std::is_invocable_v<Func, const TKey &, TCell &>;
+        using ReturnType =
+            typename std::conditional_t<with_key, std::invoke_result<Func, const TKey &, TCell &>, std::invoke_result<Func, TCell &>>::type;
+        static constexpr bool ret_bool = std::is_same_v<bool, ReturnType>;
+
+        if (!self.buf || self.empty())
+            return false;
+
+        if (pos == self.startPos())
+        {
+            pos = self.buf;
+            if (self.hasZero())
+            {
+                if constexpr (with_key)
+                {
+                    if constexpr (ret_bool)
+                    {
+                        if (func(self.zeroValue()->getKey(), *self.zeroValue()))
```

Comment:
Two embedded constexprs are a bit too much -- a plain for loop shouldn't look so complex. I'd write separate methods for different Func signatures (we already have forEachValue/forEachMapped). Probably we can have a next() method that is like operator++ from iterator, to seamlessly advance over zero key, although I'm not sure it's worth it.

Reply:
but that means a lot of duplicated code. I'll try if I can find a better way to decrease the if constexpr usage

Reply:
Duplication is OK if it increases readability -- when I see a for loop, I don't have to think, I just know what it is. But when I see multiple nested constexpr, this makes me pause for a moment. I think we should write the code in the most dull and boring way, so that it's just immediately obvious without thinking. This way we can conserve the mental energy for the problem we are solving.

Reply:
I agree with that "mental pause issue", but duplication requires additional efforts of code synchronization. I'll see what I can do.